scalyr_agent/copying_manager/checkpoints.py

```diff
@@ -0,0 +1,122 @@
+
+Since the log copying activity is distributed among worker sessions, those sessions have their own checkpoint files.
+The overall state of the particular worker session is periodically written to the checkpoint file
+named "checkpoints-worker-<WORKER_SESSION_ID>.json". This file contains the checkpoint states for every log file processor
+in the worker session.
+
+Besides that file, there is also the "active-checkpoints" file. This file contains only checkpoint states for the log processors
+that have been active during the current worker session iteration. This is needed to save the latest progress as soon as
+possible to minimize the chance of data loss if something unexpected happens.
+
+Before the agent starts, it looks for all checkpoint files that may be left from the previous run, reads and writes them
+into a "consolidated" checkpoint file named "checkpoints.json" so, after that, worker sessions can start writing to their
+checkpoint files and their previous progress won't be lost. NOTE: The consolidated file is also created on the agent stop
+but that is not so reliable, since the agent may be stopped in non-gracefully manner.
+"""
+
+
+def read_checkpoint_state_from_file(
+    file_path, logger
+):  # type: (six.text_type, AgentLogger) -> Optional[Dict]
+    """
+    Read checkpoint file from the given path and handle some basic error, if occurred.
+    """
+
+    if not os.path.exists(file_path):
+        return None
+
+    # noinspection PyBroadException
+    try:
+        checkpoints = scalyr_util.read_file_as_json(file_path, strict_utf8=True)
+    except Exception:
+        logger.exception("Cannot read the checkpoint file {0}.".format(file_path))
+        return None
+
+    # the data in the file was somehow corrupted so it can not be read as dict.
+    if not isinstance(checkpoints, dict):
+        logger.error("The checkpoint file data has to be de-serialized into dict.")
+        return None
+
+    return checkpoints
+
+
+def update_checkpoint_state_in_file(
+    checkpoints_to_save, file_path, current_time, config, logger
+):
+    # type: (Dict, six.text_type, float, Configuration, AgentLogger) -> None
+    """
+    Updates the content of the file with the checkpoint states from the given log processors.
+    Initially, it reads the checkpoints from the file(if exists) and removes all stale checkpoint states. It is better
+    to remove stale checkpoint state to prevent the checkpoint file growth.
+
+    :param checkpoints_to_save: The checkpoint states.
+    :param file_path: The checkpoint file path.
+    """
+
+    checkpoints = {}
+
+    # read existing checkpoint files (if presented).
+    file_data = read_checkpoint_state_from_file(file_path, logger)
+
+    if file_data:
+        checkpoints = file_data["checkpoints"]
+
+    # remove stale checkpoint states.
+    for path, state in list(checkpoints.items()):
+        state_timestamp = state.get("time")
+        if state_timestamp is None:
```

Comment:
Add a comment explaining what this line is for.  i.e., just to be safe, we will keep checkpoints if they do not have a timestamp.